compiler/noirc_frontend/src/hir_def/expr.rs

```diff
@@ -354,3 +355,156 @@ pub struct HirLambda {
+        Self { field, is_negative }
+    }
+}
+
+impl std::ops::Neg for SignedField {
+    type Output = Self;
+
+    fn neg(mut self) -> Self::Output {
+        self.is_negative = !self.is_negative;
+        self
+    }
+}
+
+impl std::cmp::PartialOrd for SignedField {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        if self.is_negative != other.is_negative {
+            if self.is_negative {
+                return Some(std::cmp::Ordering::Less);
+            } else {
+                return Some(std::cmp::Ordering::Greater);
+            }
+        }
+        self.field.partial_cmp(&other.field)
+    }
+}
+
+impl std::fmt::Display for SignedField {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.is_negative {
+            write!(f, "-")?;
+        }
+        write!(f, "{}", self.field)
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum Constructor {
+    True,
+    False,
+    Unit,
+    Int(SignedField),
+    Tuple(Vec<Type>),
+    Variant(Type, usize),
+    Range(SignedField, SignedField),
+}
+
+impl Constructor {
+    pub fn variant_index(&self) -> usize {
+        match self {
+            Constructor::False
+            | Constructor::Int(_)
+            | Constructor::Unit
+            | Constructor::Tuple(_)
+            | Constructor::Range(_, _) => 0,
+            Constructor::True => 1,
+            Constructor::Variant(_, index) => *index,
+        }
+    }
+
+    /// True if this constructor constructs an enum value.
+    /// Enums contain a tag value and often have values to
+    /// unpack for each different variant index.
+    pub fn is_enum(&self) -> bool {
+        match self {
+            Constructor::Variant(typ, _) => match typ.follow_bindings_shallow().as_ref() {
+                Type::DataType(def, _) => def.borrow().is_enum(),
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    /// True if this constructor constructs a tuple or struct value.
+    /// Tuples or structs will still have values to unpack but do not
+    /// store a tag value internally.
+    pub fn is_tuple_or_struct(&self) -> bool {
+        match self {
+            Constructor::Tuple(_) => true,
+            Constructor::Variant(typ, _) => match typ.follow_bindings_shallow().as_ref() {
+                Type::DataType(def, _) => def.borrow().is_struct(),
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+}
+
+impl std::fmt::Display for Constructor {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Constructor::True => write!(f, "true"),
+            Constructor::False => write!(f, "false"),
+            Constructor::Unit => write!(f, "()"),
+            Constructor::Int(x) => write!(f, "{x}"),
+            Constructor::Tuple(_) => Ok(()), // args are printed after already
```

Comment:
What does this mean?

Reply:
This is just the constructor itself, so the function value. We print the arguments of the constructor e.g. `(x, y)` afterwards anyway and in that case it already matches a tuples syntax so we don't want to print an extra `()` or similar before it since that wouldn't be correct. I'll clarify the comment.